src/core/domain/swap.ts

```diff
@@ -170,63 +160,50 @@ export class Swap extends SwapPure {
-        tx = router.swapExactTokensForKLAY(
-          [props.amountIn.asStr, props.amountOutMin.asStr, path, address, deadline],
-          baseOverrides,
-        )
-
-        break
-      }
-      case 'exact-eth-for-tokens': {
-        tx = router.swapExactKLAYForTokens([props.amountOutMin.asStr, path, address, deadline], {
-          ...baseOverrides,
-          value: props.amountIn,
-        })
-
-        break
-      }
-      case 'eth-for-exact-tokens': {
-        tx = router.swapKLAYForExactTokens([props.amountOut.asStr, path, address, deadline], {
-          ...baseOverrides,
-          value: props.amountInMax,
-        })
-
-        break
-      }
-      case 'tokens-for-exact-eth': {
-        tx = router.swapTokensForExactKLAY(
-          [props.amountOut.asStr, props.amountInMax.asStr, path, address, deadline],
-          baseOverrides,
-        )
-
-        break
-      }
-
-      default: {
-        const badProps: never = props
-        throw new Error(`Bad props: ${String(badProps)}`)
-      }
-    }
+    const isInputTokenNative = isNativeToken(props.trade.route.input.address)
+    const isOutputTokenNative = isNativeToken(props.trade.route.output.address)
+
+    const tx: TransactionObject<unknown> = match([isInputTokenNative, isOutputTokenNative] as [boolean, boolean])
+      .with([false, false], () =>
+        match(props)
+          .with({ mode: 'exact-in' }, ({ amountIn, amountOutMin }) =>
+            router.swapExactTokensForTokens([amountIn.asStr, amountOutMin.asStr, path, address, deadline]),
+          )
+          .with({ mode: 'exact-out' }, ({ amountInMax, amountOut }) =>
+            router.swapTokensForExactTokens([amountOut.asStr, amountInMax.asStr, path, address, deadline]),
+          )
+          .exhaustive(),
+      )
+      .with([false, true], () =>
+        match(props)
+          .with({ mode: 'exact-in' }, ({ amountIn, amountOutMin }) =>
+            router.swapExactTokensForKLAY([amountIn.asStr, amountOutMin.asStr, path, address, deadline]),
+          )
+          .with({ mode: 'exact-out' }, ({ amountInMax, amountOut }) =>
+            router.swapTokensForExactKLAY([amountOut.asStr, amountInMax.asStr, path, address, deadline]),
+          )
+          .exhaustive(),
+      )
+      .with([true, false], () =>
+        match(props)
+          .with({ mode: 'exact-in' }, ({ amountIn, amountOutMin }) =>
+            router.swapExactKLAYForTokens([amountOutMin.asStr, path, address, deadline], {
+              ...baseOverrides,
+              value: amountIn,
+            }),
+          )
+          .with({ mode: 'exact-out' }, ({ amountInMax, amountOut }) =>
+            router.swapKLAYForExactTokens([amountOut.asStr, path, address, deadline], {
+              ...baseOverrides,
+              value: amountInMax,
+            }),
+          )
+          .exhaustive(),
+      )
+      .with([true, true], () => {
+        throw new Error('unreachable')
+      })
+      .exhaustive()
```

Comment:
@SergeyPoslavskiy, you would help **a lot** if you check whether all arguments here are passed correctly.